### ISSUE TITLE
[false-positive] mkdir race conditions with literal and/or

### ISSUE THREAD

**Issue by marcelloh** [kalessil/phpinspectionsea#23]
`if (!@mkdir($dir, 0777, true) and !is_dir($dir))`
which is the same as
`if (!@mkdir($dir, 0777, true) && !is_dir($dir))`

but I use the more readable `and`.

But it blames me for the wrong reasons :-(

**Comment by kalessil** [comment]
Thanks for reporting, @marcelloh .

False-positive is confirmed, I will ping you when it fixed.

**Comment by dryabov** [comment]
@kalessil I have false-positive for following code:
```php
    public function makeDir($path, $chmod = 0777)
    {
        return is_dir($path) || @mkdir($path, $chmod, true) || is_dir($path);
    }
```

PS. And there is a conflict with "This condition is duplicated in other if/elseif branch" for similar code:
```php
        if (!is_dir($destDir) && !@mkdir($destDir, 0755, true) && !is_dir($destDir)) {
            trigger_error("Cannot create directory $destDir");
        }
```

**Comment by marcelloh** [comment]
somehow when the order is the "adviced order"
`
        if (!@mkdir($dir, 0777, true) and !is_dir($dir))
`
it still throws a warning or error or something: mkdir(): File exists

But when I change the order into:
`
if (!is_dir($dir) and !mkdir($dir, 0777, true))
`
all is well, without any warning

**Comment by funivan** [comment]
@dryabov Why you are using `!is_dir` before `mkdir` ? 
mkdir is always executed because you are using `&&` operator . So there is no point to check `is_dir` twice.

**Comment by kalessil** [comment]
Fixed in 2.3.10, which I releasing today evening.
